Initial take on Histogram._fillnd

pavoljuhas opened the pull request:
@chrisvam, @ordirules, @ericdill - please review and let me know if you feel it is good to merge.

pavoljuhas committed:
Implement Histogram._fillnd.

Also add consistency test for 1D and 2D histograms.

pavoljuhas committed:
Compare timings of Histogram._fill1d and _fillnd.

Both are run on the same data.

ericdill commented on skxray/core/accumulators/histogram.pyx:
```diff
@@ -121,7 +129,7 @@ class Histogram:
         else:
             # do the generalized ND histogram
-            raise NotImplementedError()
+            self._fillnd(coords, weights)
```
This makes me really happy :)  Hooray for nD histogramming!

ericdill commented:
Tests for 1d and 2d are passing locally. Worth testing the higher dimensionality histogrmaming?

jrmlhermitte commented on skxray/core/accumulators/histogram.pyx:
```diff
@@ -164,28 +172,72 @@ class Histogram:
+        cdef int afloat_stride[MAX_DIMENSIONS]
+        cdef int afloat_count = 0
+        istrides = np.asarray(self._values.strides, dtype=np.int32)
+        istrides //= self._values.itemsize
```
Very neat way to get the strides!

jrmlhermitte commented on skxray/core/accumulators/histogram.pyx:
```diff
@@ -164,28 +172,72 @@ class Histogram:
+        # numerical data type
+        numtypes = [np.dtype(int), np.dtype(float)]
+        numtypeindex = {tp : i for i, tp in enumerate(numtypes)}
+        ctypeindices = [numtypeindex.get(c.dtype, MAX_DIMENSIONS)
```
should this be len(numtypeindex)+1 and not MAX_DIMENSIONS?

ordirules commented:
This is very elegant, good job Pavol! You have taught me a lot of neat tricks in python (which I assumed you also just taught yourself in the past 2 days or so ;-) )

I took a quick look now. I think benchmarking would be a good idea @ericdill  no? I will also try running it on my 2D simulation (basically, go from a 1D detector to 2D detector and do the same sort of thing). Moving via striding is always tricky and often leads to small issues.

pavoljuhas committed:
Improve coordinates ordering by type in fillnd.

Use a constant, very large sort key for unknown types.

jrmlhermitte committed:
Added extra test that fillnd fails on. Will need to fix

pavoljuhas committed:
Fix check of Histogram arguments and PEP8 issues.

Picked updates in ordirules/scikit-xray@0e77be5b5a127aa8a6d519585a7ed4fb09942a1c.

pavoljuhas committed:
Fix order of data strides in ND histogram.

In ND histogram the coordinate arrays are looped in the order given by
their numerical types.  This avoids the need to check numerical types
in the innermost loop.  The index strides for the target ND array
must be in a consistent order.

pavoljuhas committed:
Use x when looping over inputs in ND histogram.

No change in code function.  This closes #8.
